enhancements/implicit-catalog-versioning.md

```diff
@@ -0,0 +1,152 @@
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+    replaces: etcdoperator.v0.9.0
+```
+
+The `replaces` field is the source for the update graph for each install channel in a given operator. When the `operator-registry` builds the update graph, it actually walks through each CSV in a given Operator and, for each channel, constructs a chain of CSVs that define the update path. Instead, we can just define our spec without the replaces field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+```
+
+We would use the fact that 0.9.2 > 0.9.0 to implicitly assume that 0.9.0 is replaced by 0.9.2 through the rules of semantic version.
+
+To accomplish this:
+- Make `spec.version` a required field in the operator-registry project
+- Update the `operator-registry` to walk through the set of CSVs and construct the replacement chain by reading the `version` field instead of `replaces`. This will be done by modifying the Channel tables to point to next ordered semantic versions when defining the graph
+- Ignore `skips` field entirely when building the graph
+- Update OLM to use the query for the replacement rather than the current value in the existing CSV
+
+Note that these requirements will only be the case when generating update graphs with the latest version of the operator-registry's `opm registry` commands. 
+
+### Mode
+
+First, a quick note about `skipRange`: The `skipRange` annotation was originally added to allow an explicit definition of version ranges that can be skipped over -- generally this was, done in order to allow patch versions of released operators to upgrade directly. However, given that version is becoming a first class citizen, some users may expect that patch version to work in that fashion in general.
+
+In order to account for this possibility, we will explicitly ask the user when adding bundles which method they want to use to upgrade. To handle this, `opm index` and `opm bundle` will add a new flag `--mode` that enumerates to three different options:
+
+`opm index add quay.io/namespace/my-operator --mode=semver`
+
+In this case, we will use the semantic version to define the graph. Simple comparison drives the list, so 1.1.0 is replaced by 1.1.1 is replaced by 1.1.2 is replaced by 1.2.0. ex:
+
+`opm index add quay.io/namespace/my-operator --mode=semver-skippatch`
+
+In this case, we will use the semantic version to define the graph while skipping directly to the latest patch version. Here 1.1.0 is replaced by 1.1.1 until 1.1.2 is added. At that point, 1.1.2 skips directly over 1.1.1 when upgrading 1.1.0. But 1.1.2 is replaced by 1.2.0.
```

Comment:
I would expect the default `semver` to be the one that skips z's, and actually y's as well. So I would think I would have options to affect stepping through both patch and minor versions as well. WDYT about `semver`, `semver-walk-y`, `semver-walk-z`, `semver-walk-yz`?

Reply:
of course, your design is flexible enough to add as many as we want later on.

Reply:
I am less convinced at the moment of wanting to make it easy to support walking over `y`. I would like to note that this is currently not an optional parameter -- there is no default. I believe we should just start with these options and based on community feedback eventually select a default. Of course, if there is desire for it this does pretty easily open up to use this option for minor version stepping as well.

Reply:
from a slack discussion:

> that’s fair — I guess it comes down to whether our ideas of stepping though each patch version of an operator is what users expect. It’s what OLM does today, but most redhat operators are packaged to upgrade directly to the latest patch.
>
> In my head I was trying to equate API compatibility with upgradeability
and that patch=bugfix => non-latest patches contain bugs => don’t want to install bugged content